sdk/eventhubs/azure-messaging-eventhubs/src/samples/java/com/azure/messaging/eventhubs/PublishStreamOfEvents.java

```diff
@@ -0,0 +1,234 @@
+        private final AtomicBoolean isDisposed = new AtomicBoolean();
+        private final AtomicReference<EventDataBatch> currentBatch = new AtomicReference<>();
+        private final EventHubProducerAsyncClient producer;
+        private final CreateBatchOptions batchOptions;
+        private final Duration windowDuration;
+
+        /**
+         * Creates a new instance of {@link CustomPublisher}.
+         *
+         * @param connectionString Connection string for the specific Event Hub.
+         * @param windowDuration Intervals to check for an available {@link EventDataBatch} to send.
+         * @param batchOptions Options to use when creating the {@link EventDataBatch}. If {@code null}, the default
+         * batch options are used.
+         */
+        CustomPublisher(String connectionString, Duration windowDuration, CreateBatchOptions batchOptions) {
+            producer = new EventHubClientBuilder()
+                .connectionString(connectionString)
+                .buildAsyncProducerClient();
+
+            this.batchOptions = batchOptions != null ? batchOptions : new CreateBatchOptions();
+            this.windowDuration = Objects.requireNonNull(windowDuration, "'windowDuration' cannot be null.");
+        }
+
+        /**
+         * Subscribes to a stream of events and publishes event batches when:
+         *
+         * <ul>
+         * <li>The {@link EventDataBatch} is full.</li>
+         * <li>Timeout window has elapsed and there is a batch.</li>
+         * </ul>
+         *
+         * @param events Events to publish to the service.
+         *
+         * @return Mono that completes when all the events have been published.
+         */
+        Mono<Void> publish(Flux<EventData> events) {
+            final Flux<EventDataBatch> fullBatchFlux = Flux.<EventDataBatch>create(sink -> {
+                events.subscribe(event -> {
+                    // For each event, tries to add it to the current batch. If the current batch is full, then we emit
+                    // that. Otherwise an empty Mono is emitted.
+                    // Blocking at the very end of the method to ensure that the event is added to a batch before
+                    // requesting another event.
+                    getOrCreateBatch()
+                        .flatMap(batch -> batch.tryAdd(event) ? Mono.empty() : Mono.just(batch))
+                        .flatMap(fullBatch -> {
+                            sink.next(fullBatch);
+
+                            return createBatch().map(newBatch -> {
+                                if (!newBatch.tryAdd(event)) {
+                                    sink.error(new IllegalArgumentException(String.format(
+                                        "Event is too large for an empty batch. Max size: %s. Event: %s",
+                                        newBatch.getMaxSizeInBytes(), event.getBodyAsString())));
```

Comment:
Instead of printing the event body, it might be better to show how to get the event size. This will show the comparison of the max size allowed and how much bigger the event size is.

Reply:
We currently don't persist this information in EventData and it is just calculated. It would be nice to know. I'll create an issue to can add this.